### PR TITLE
Idempotency caches over HTTP for OPEN, BID, and UPDATE

### PR DESCRIPTION
**Pull request by RevCBH**

### CODE REVIEW

File: lib/wallet/http.js [hunk 1]
```diff
@@ -1074,12 +1074,20 @@ class HTTP extends Server {
...
       assert(name, 'Name is required.');
       assert(broadcast ? sign : true, 'Must sign when broadcasting.');
 
+      if (idempotencyKey) {
+        assert(broadcast && sign, 'Must sign and broadcast if using idempotency cache')
+      }
+
       const options = TransactionOptions.fromValidator(valid);
-      const mtx = await req.wallet.createOpen(name, force, options);
+      const [mtx, fromCache] = await req.wallet.withOpenCache(idempotencyKey, () =>
```
Comment (turbomaze) [inline lib/wallet/http.js:1087]: style nit: implicit returns should only be used for single line functions, too easy to miss while skimming, use
```javascript
...withOpenCache(idempotencyKey, () => {
  return req.wallet.createOpen(name, force, options);
});
```

File: lib/wallet/http.js [hunk 2]
```diff
@@ -1074,12 +1074,20 @@ class HTTP extends Server {
...
+        req.wallet.createOpen(name, force, options)
```
Comment (turbomaze) [inline lib/wallet/http.js:1088]: Linter doesn't seem to be running, can you make sure eslint goes thru this to add semicolons everywhere etc

File: lib/wallet/wallet.js [hunk 3]
```diff
@@ -1736,16 +1736,9 @@ class Wallet extends EventEmitter {
...
-        const result = this.sendOpenResults.get(idempotencyKey);
-        return [result, true];
-      }
-
-      const result = await this._sendOpen(name, force, options);
-      if (idempotencyKey) {
-        this.sendOpenResults.set(idempotencyKey, result);
-      }
-      return [result, false];
+      return await this.withOpenCache(idempotencyKey,
```
Comment (turbomaze) [inline lib/wallet/wallet.js:1739]: Nice

File: test/wallet-test.js [hunk 4]
```diff
@@ -2494,4 +2495,25 @@ describe('Wallet', function() {
       assert.equal(claim.covenant.isClaim(), true);
     });
   });
+
+  describe('Idempotency caching', () => {
+    it.only('should have a generic mechanism for hitting a cache', async () => {
```
Comment (turbomaze) [inline test/wallet-test.js:2500]: remove only

File: lib/wallet/rpc.js [hunk 5]
```diff
@@ -2141,7 +2145,8 @@ class RPC extends RPCBase {
     if (bid == null || value == null)
       throw new RPCError(errs.TYPE_ERROR, 'Invalid values.');
 
-    if (bid > value) throw new RPCError(errs.TYPE_ERROR, 'Invalid bid.');
+    if (bid > value)
+throw new RPCError(errs.TYPE_ERROR, 'Invalid bid.');
```
Comment (turbomaze) [inline lib/wallet/rpc.js]: This indentation seems incorrect

File: lib/wallet/wallet.js [hunk 6]
```diff
@@ -1868,27 +1861,65 @@ class Wallet extends EventEmitter {
...
+    if (idempotencyKey && cache.has(idempotencyKey)) {
+      return {result: cache.get(idempotencyKey), fromCache: true};
+    }
 
+    const result = await callback();
+    if (idempotencyKey) {
+      cache.set(idempotencyKey, result);
+    }
+
+    return [result, false];
```
Comment (turbomaze) [inline lib/wallet/wallet.js:1881]: different return signature

File: lib/wallet/http.js [hunk 7]
```diff
@@ -1200,7 +1238,14 @@ class HTTP extends Server {
       }
 
       const options = TransactionOptions.fromValidator(valid);
-      const mtx = await req.wallet.createUpdate(name, resource, options);
+
+      const [mtx, fromCache] = await req.wallet.withBidCache(
```
Comment (turbomaze) [inline lib/wallet/http.js:1242]: grep for all this usage the signature has changed

**Comment by RevCBH** [comment]
Verified over rpc and http

**Pull request merged**